### PR TITLE
feat: concurrent request support added

### PR DESCRIPTION
**Pull request by erayarslan**

### CODE REVIEW

**Review by Abdulsametileri — APPROVED** [review]

File: helper/reader.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package helper
+
+import (
+	"io"
+)
+
+type MultiDimByteReader struct {
```
Comment (Abdulsametileri) [inline helper/reader.go:7]: I think, we need to add tests for reader and utils

File: helper/reader.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	s                          [][]byte
+	currentSliceIndex          int
```
Comment (Abdulsametileri) [inline helper/reader.go:9]: What does `current` mean? We can also add comment statements on the top of related fields. This comments should be include what and why maybe.

File: helper/utils.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package helper
+
+func Flatten(s [][]byte) []byte {
```
Comment (Abdulsametileri) [inline helper/utils.go:3]:
```go
func Flatten(collection [][]byte) []byte {
	totalLen := 0
	for i := range collection {
		totalLen += len(collection[i])
	}

	result := make([]T, 0, totalLen)
	for i := range collection {
		result = append(result, collection[i]...)
	}

	return result
}
```

Maybe we can rearrange Flatten operation like this? Or We can do some benchmark

File: elasticsearch/bulk/bulk.go [hunk 4]
```diff
@@ -190,25 +216,49 @@ func (b *Bulk) flushMessages() {
 		}
 		b.batchTicker.Reset(b.batchTickerDuration)
 		b.batch = b.batch[:0]
+		b.batchKeys = make(map[string]int, b.batchSizeLimit)
+		b.batchIndex = 0
```
Comment (Abdulsametileri) [inline elasticsearch/bulk/bulk.go:220]: 0 is meaningful in terms of index. Is it really necessary to assign this value to 0? If not we can assign it with like -1 :)

File: elasticsearch/bulk/bulk.go [hunk 5]
```diff
@@ -190,25 +216,49 @@ func (b *Bulk) flushMessages() {
...
 		b.batchSize = 0
+		b.batchByteSize = 0
 	}
 
 	b.dcpCheckpointCommit()
 }
 
+func (b *Bulk) request(concurrentRequestIndex int, batch [][]byte) func() error {
```
Comment (Abdulsametileri) [inline elasticsearch/bulk/bulk.go:228]: I think `requestFunc` is more clear instead of `request`

File: elasticsearch/bulk/bulk.go [hunk 6]
```diff
@@ -21,23 +25,27 @@ import (
...
+	collectionIndexMapping map[string]string
+	batchKeys              map[string]int
 	dcpCheckpointCommit    func()
 	batchTicker            *time.Ticker
 	isClosed               chan bool
 	actionCh               chan document.ESActionDocument
 	esClient               *elasticsearch.Client
-	metric                 *Metric
-	collectionIndexMapping map[string]string
+	batch                  [][]byte
```
Comment (Abdulsametileri) [inline elasticsearch/bulk/bulk.go:38]: You can also extract `[][]byte` into the custom type alias like

type BatchData [][]byte etc.

File: elasticsearch/bulk/bulk.go (hunk 4 above)
Comment (erayarslan) [reply elasticsearch/bulk/bulk.go:220]: It's necessary because my goal is to get it back on index 0.

File: helper/utils.go (hunk 3 above)
Comment (erayarslan) [reply helper/utils.go:3]: This method is no longer needed, so it will be deleted.

File: helper/reader.go (hunk 1 above)
Comment (erayarslan) [reply helper/reader.go]: Done with love.

File: elasticsearch/bulk/bulk.go (hunk 6 above)
Comment (erayarslan) [reply elasticsearch/bulk/bulk.go:38]: It is not something that is reused. I don't want to make unnecessary imports.

**Pull request merged**